### PR TITLE
Magfield

### PR DESCRIPTION
**Pull request by whokion**
Add magnetic field propagator/integrator (#49)
- first (minimal) implementation of magnetic field propagation and integration based on Geant4
- tests with a uniform magnetic field

It still need a lot of refinement and extension, but better to be integrated as a baseline.

### CODE REVIEW

**Comment by whokion** [comment]
@sethrj can you help to spot where the continuous-integration error comes from? - I cannot clearly see it from the Jenkins log.  Thanks.

**Comment by sethrj** [comment]
> @sethrj can you help to spot where the continuous-integration error comes from? - I cannot clearly see it from the Jenkins log. Thanks.

Looks like you can see it in [the full log](https://cloud.cees.ornl.gov/jenkins-ci/blue/rest/organizations/jenkins/pipelines/Celeritas/branches/PR-155/runs/1/nodes/10/steps/35/log/?start=0) -- 
```
../src/field/RungeKutta.i.hh: In member function 'void celeritas::RungeKutta::ode_stepper(celeritas::real_type, celeritas::RungeKutta::ode_type, const ode_type&, celeritas::RungeKutta::ode_type&)':
../src/field/RungeKutta.i.hh:31:24: error: implicitly-declared 'constexpr celeritas::OdeArray& celeritas::OdeArray::operator=(const celeritas::OdeArray&)' is deprecated [-Werror=deprecated-copy]
   31 |     yt = y + 0.5 * h * dydx;
      |                        ^~~~
In file included from ../src/field/FieldEquation.hh:12,
                 from ../src/field/FieldStepper.hh:12,
                 from ../src/field/RungeKutta.hh:12,
                 from ../src/field/FieldIntegrator.hh:13,
                 from ../test/field/FieldIntegrator.test.cc:10:
../src/field/base/OdeArray.hh:57:5: note: because 'celeritas::OdeArray' has user-provided 'celeritas::OdeArray::OdeArray(const celeritas::OdeArray&)'
   57 |     OdeArray(OdeArray const &v)
      |     ^~~~~~~~
```

**Comment by sethrj** [comment]
I don't think I'll be able to review this until the weekend unfortunately :( with the CHEP paper and other deadlines.

**Comment by whokion** [comment]
No problem.  For the mean time, I will try to address the error.

**Comment by sethrj** [comment]
@whokion I am taking a peek here and there at this merge request. It looks like clang-format didn't get applied. Could you run `clang-format -i src/field/* test/field/*`, commit, and push?

**Comment by whokion** [comment]
Applied clang-format again - I need to understand why the clang-format setup sometimes does not work.

File: src/field/FieldIntegrator.i.hh [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    ode_type dydx;
+    ode_rhs(y, dydx);
+
+    for (CELER_MAYBE_UNUSED int i : celeritas::range(shared_.max_nsteps))
+    {
+        // Always start from the initial point
+        yend = y;
+        real_type dchord_step;
+
+        dyerr = quick_advance(h, yend, dydx, dchord_step);
```
Comment (sethrj) [inline src/field/FieldIntegrator.i.hh:81]: Prefix all member function calls with `this->` for improved readability.

File: src/field/FieldParamsPointers.hh [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    real_type delta_chord;
+
+    //! accuracy of intersection of a boundary crossing  = 0.0001 mm
+    real_type delta_intersection;
+
+    //! the relative error = 1.0e-5
+    real_type epsilon_step;
+
+    //! min step size = 1.0e-5
+    real_type minimun_step;
```
Comment (sethrj) [inline src/field/FieldParamsPointers.hh:34]: Typo in this variable name -- is this variable used, or is the typo elsewhere in the code?

File: src/field/FieldStepper.hh [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    inline CELER_FUNCTION FieldStepper(FieldEquation& equation);
+
+    // Commom methods
+    CELER_FUNCTION real_type stepper(real_type       h,
+                                     const ode_type  y,
+                                     const ode_type& dydx,
+                                     ode_type&       yout);
+
+    CELER_FUNCTION real_type distance_chord(const ode_type y,
+                                            const ode_type yout);
```
Comment (sethrj) [inline src/field/FieldStepper.hh:40]: A lot of these functions should be `const ode_type&`.

File: src/field/FieldStepper.hh [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Brief class description.
+ *
+ * This is the base class of field steppers
+ */
+template<typename T>
+class FieldStepper
```
Comment (sethrj) [inline src/field/FieldStepper.hh:25]: What's the template parameter here? Also the class documentation (and other places where "Brief class description." appear) need to be updated.

File: src/field/MagField.hh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    //! Type aliases
+    using field_value = Real3;
+    //@}
+
+  public:
+    // Construct from a uniform field
+    CELER_FUNCTION MagField(const field_value& value);
+
+    // return a uniform magnetic field value
+    CELER_FUNCTION field_value operator()();
```
Comment (sethrj) [inline src/field/MagField.hh:35]: Should this eventually be a function of position and/or volume? We could put those in as a placeholder showing the eventual intent, and just not use them for the uniform case.

File: src/field/base/OdeArray.hh [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    CELER_FUNCTION
+    OdeArray(const real_type c)
+    {
+        for (auto i : range(ode_dim))
+            v_[i] = c;
+    }
+
+    CELER_FUNCTION
+    OdeArray(const ode_type v)
```
Comment (sethrj) [inline src/field/base/OdeArray.hh:51]: Does the ODE array ever get constructed from an ODE type? Wouldn't it be better to construct from `const Real3& position, const Real3& momentum`? Did you consider using that for storage under the hood as well? It could make some of your operations easier as well as allowing const-reference access to the data as opposed to copy-by-value.

File: src/field/base/OdeArray.hh [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    {                                                \
+        for (auto i : range(ode_dim))                \
+            v_[i] OPERATOR c;                        \
+        return *this;                                \
+    }
+    INPLACE_BINARY_OP(+=)
+    INPLACE_BINARY_OP(-=)
+    INPLACE_BINARY_OP(*=)
+    INPLACE_BINARY_OP(/=)
+#undef INPLACE_BINARY_OP
```
Comment (sethrj) [inline src/field/base/OdeArray.hh:112]: Do all these operators make sense? Are you really adding the same scalar to both the position and momentum?

File: src/field/base/OdeArray.hh [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    // Constructors
+    CELER_FUNCTION
+    OdeArray()
+    {
+        for (auto i : range(ode_dim))
+            v_[i] = 0;
+    }
+
+    CELER_FUNCTION
+    OdeArray(const real_type c)
```
Comment (sethrj) [inline src/field/base/OdeArray.hh:44]: This should definitely be a explicit constructor -- otherwise a function that expects an OdeArray could get unintentionally passed an integer/double.

I'm actually not sure why this is needed at all -- in what circumstances do the position and momentum have the same numerical value?

**Review by sethrj — CHANGES_REQUESTED** [review]
Thanks Soon! There are a couple of readability issues that would help my review before the next iteration -- specifically, the use of mutable references to pass data around (which makes it very hard to tell from the calling code what is being modified. Could you try to address those changes and then I'll work on another iteration of the review?

File: src/field/FieldParamsPointers.hh (hunk 2 above)
Comment (whokion) [reply src/field/FieldParamsPointers.hh:34]: The typo is fixed in all places where it is actually set or used.

File: src/field/MagField.hh (hunk 5 above)
Comment (whokion) [reply src/field/MagField.hh:35]: Yes, both constructor and the field retrieving method should be extended in the next round (underway).  Also this class will be served a base for the user magnetic field class.  I just added a dummy method with a position argument as suggested for this MR.

File: src/field/base/OdeArray.hh (hunk 6 above)
Comment (whokion) [reply src/field/base/OdeArray.hh:51]: I used this constructor in field tests (to initialize field states).  I was thinking a single interface to initialize the state when ode_type is extended later including time, spin and so on.  Nevertheless, this constructor will be never used as
the field state will be initialized based on geo/particle track view in a real workflow.  Anyway, the suggested one is better in the current ontext (6 variables with position and momentum), but the constructor will be dropped as there is another way to initialize the state in test routines.

File: src/field/base/OdeArray.hh (hunk 7 above)
Comment (whokion) [reply src/field/base/OdeArray.hh:112]: You are right.  Drop + and - for scalar.  Also did same for ODE_BINARY_OP.

File: src/field/base/OdeArray.hh (hunk 8 above)
Comment (whokion) [reply src/field/base/OdeArray.hh:44]: I was thinking for the case that the whole array is initialized a large value.  Again it is not essential and the constructor is dropped.

File: src/field/FieldStepper.hh (hunk 4 above)
Comment (whokion) [reply src/field/FieldStepper.hh:25]: Added "\tparam T A derived class.".  Will add more descriptions for other classes.

File: src/field/FieldIntegrator.i.hh (hunk 1 above)
Comment (whokion) [reply src/field/FieldIntegrator.i.hh:81]: Done. Also in other classes.

File: src/field/FieldIntegrator.i.hh [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+        // Exit if the distance to the chord is small than the reference
+        if (dchord_step <= shared_.delta_chord)
+            break;
+
+        // try a reduced step size, but not more than a factor 2
+        h *= std::fmax(sqrt(shared_.delta_chord / dchord_step), 0.5);
+    }
+    //! XXX: loop check for rare failed cases?
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldIntegrator.i.hh:90]: Use do-while in all possible cases.

File: src/field/FieldStepper.hh (hunk 4 above)
Comment (sethrj) [reply src/field/FieldStepper.hh:25]: I still don't understand what that means here -- is this a "curiously recursive template pattern"? What interface is T expected to have?

File: src/field/FieldIntegrator.i.hh [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+    return h;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Adaptive step control
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldIntegrator.i.hh:97]: Added  separators and  more function descriptions.

File: src/field/FieldStepper.hh (hunk 4 above)
Comment (whokion) [reply src/field/FieldStepper.hh:25]: Yes, it is CRTP.

**Comment by sethrj** [comment]
Thanks for the updates @whokion . I'll be working on this today.

File: src/field/FieldIntegrator.i.hh [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+       if(!converged)  {
+      	    if (CELER_UNLIKELY(--step_countdown == 0))
+	    {
+	        // Failed to converge, handle rare cases here
+	        break;
+	    }
+            // try a reduced step size, but not more than a factor 2
+            h *= std::fmax(sqrt(shared_.delta_chord / dchord_step), 0.5);
+	}
+    } while (!converged);
```
Comment (sethrj) [inline src/field/FieldIntegrator.i.hh:96]: Originally your loops were written as though the point was to count a certain
number of steps, and converging early was an afterthought.
```c++
setup();
for (int i = 0; i < max_steps; ++i)
{
    step();
    if (is_converged())
        break;
    pre_step();
}
```

After my half-baked suggestion, the loops are now written as though the point
is to converge, but there's extra complication from having to check for
convergence twice on every loop, and also if you happen to converge at the same
point the step limiter is hit, you're testing the convergence for nothing.
```c++
bool converged;
int remaining_steps = max_steps;
setup();
do
{
    step();
    converged = is_converged();
    if (!converged)
    {
        if (--remaining_steps == 0)
            break;
        pre_step();
    }
} while (!converged);
```

I think you should flip this around by refactoring the loop so that the
"before-next-step" always happens inside the step, before the next step. The
first iteration happens outside the loop, and then the remaining iterations are
inside the loop.
```c++
setup();
step();
int remaining_steps = max_steps;
while (!is_converged() && --remaining_steps > 0)
{
    pre_step();
    step();
}
```

The first version has two jumps (`break` and `for`) and two conditonals (one as
part of the `for`); the second has two jumps and three conditionals; the last
has a single conditional and a single jump.

Minimizing jumps and conditionals has long been recognized as important to
reducing code complexity and unexpected bugs, so this last version is not only
the clearest in intent but also the least likely to be buggy.

File: src/field/RungeKutta.hh [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                                    const ode_type& dydx,
+                                    ode_type&       yout);
+
+  private:
+    // The base class is a friend of this
+    friend class FieldStepper<RungeKutta>;
+
+  private:
+    ode_type dydxm;
+    ode_type dydxt;
```
Comment (sethrj) [inline src/field/RungeKutta.hh:43]: Private data needs to have trailing `underscores_`.

File: src/field/FieldStepper.hh [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    CELER_FUNCTION void ode_rhs(ode_type y, ode_type& yout);
+
+    // Static interfaces (Mandatory methods)
+    CELER_FUNCTION void ode_stepper(real_type       h,
+                                    const ode_type  y,
+                                    const ode_type& dydx,
+                                    ode_type&       yout)
+    {
+        static_cast<T*>(this)->ode_stepper(h, y, dydx, yout);
+    }
```
Comment (sethrj) [inline src/field/FieldStepper.hh:51]: I don't think you need a recursive pattern here since `T` only needs to access the equation and not any helper functions or complex interfaces. It's very difficult to tell what data is shared, and what's local, because:
- there are missing `_` indicating member data
- lots of methods aren't marked `const` when they should be
- lots of data is being passed around by mutable reference.

Please address all instances of these last three bullets and then I'll continue reviewing.

**Review by sethrj — CHANGES_REQUESTED** [review]

File: src/field/FieldEquation.i.hh [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+CELER_FUNCTION
+void FieldEquation::evaluate_rhs(const Real3    B,
+                                 const ode_type y,
+                                 ode_type&      dy) const
+{
+    /*** m d^2x/dt^2 = (q/c)(vxB), s = |v|t, y = dx/ds
+         rhs:  dx/ds = v/|v|
+               dy/ds = (q/pc)(yxB)
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.i.hh:59]: Done (with adding \vec{...} for x, v, y, B)

File: src/field/FieldEquation.i.hh [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    */
+
+    dy[0] = y[3];
+    dy[1] = y[4];
+    dy[2] = y[5];
+    dy[3] = coeffi_ * (y[4] * B[2] - y[5] * B[1]);
+    dy[4] = coeffi_ * (y[5] * B[0] - y[3] * B[2]);
+    dy[5] = coeffi_ * (y[3] * B[1] - y[4] * B[0]);
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.i.hh:67]: These were my original expressions too, but realized that pos and mon are Real3 (so, op= does not work here).  I also have not introduced modifiers for OdeArray position and momentum yet.  I am going to use vecgeom::Vector3D for three vector in all field codes in the next round of commits (if Real3 will be a just plain array as now) and use the celeritas unit consistently every where - current unit conventions are mm for pos, MeV for mon and field value in Tesla which were assumed during initialization of FieldTrackView and MagField.  The same story for the c*1e-8 which will be consistently changed by dropping the clhep-style unit conventions later.  Of course two equations do not make sense without the last line.

File: src/field/FieldIntegrator.hh [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+    CELER_FUNCTION void ode_rhs(const ode_type y, ode_type& dydx)
+    {
+        stepper_.ode_rhs(y, dydx);
+    }
+
+    // >>> COMMON PROPERTIES
+
+    //! XXX move to constants?
+    static CELER_CONSTEXPR_FUNCTION real_type permillion() { return 1e-6; }
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldIntegrator.hh:74]: Again, permillion was adopted from clhep constants.  Changed as suggested.

File: src/field/FieldEquation.hh [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+/*!
+ * The FieldEquation evaluates the righ hand side of the Lorentz equation
+ * for a given magnetic field value.
+ */
+class FieldEquation
+{
+  public:
+    //@{
+    //! Type aliases
+    using ode_type = OdeArray;
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.hh:26]: OdeArray will be a template in the next extension (i.e., extension to include components of time, spin and etc. for supporting more general types of equations).  ode_type = OdeArray<real_type, N> where N could be 6, 7, 10, ... 
but can be used persistently in steppers (and integrators).  So, I leave it like this for now, but will think about the better definition.  Any suggestion?

File: src/field/FieldEquation.hh [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+  public:
+    // Construct with a magnetic field
+    CELER_FUNCTION FieldEquation(MagField& field);
+
+    // set charge if q != -eplus
+    CELER_FUNCTION void set_charge(real_type q);
+
+    // evaluate the right hand side
+    // XXX: add the position arguement for a non-uniform field
+    CELER_FUNCTION void operator()(const ode_type y, ode_type& dydx) const;
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.hh:38]: Very good suggestion!  Done.

File: src/field/FieldIntegrator.hh [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    using ode_type = OdeArray;
+
+  public:
+    // Construct with shared data and the stepper
+    inline CELER_FUNCTION
+    FieldIntegrator(const FieldParamsPointers& shared, RungeKutta& stepper);
+
+    // Interfaces from G4ChordFinder
+    CELER_FUNCTION real_type advance_chord_limited(real_type step_trial,
+                                                   ode_type& y);
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldIntegrator.hh:40]: Yes, I have not done necessary changes yet - I just pushed what I updated yesterday. Will Do.

[the remark this answers is not in the document]
Comment (sethrj) [reply src/field/FieldIntegrator.hh:40]: Sorry, I didn't realize that. I saw a response that I thought indicated it was taken care of, apologies for the misunderstanding.

File: src/field/FieldEquation.hh (hunk 17 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/field/FieldEquation.hh:26]: Aha, good point. I understand why the spin vector would changes as the particle changes direction through the  field, but why would the time be affected? Can't the time just be updated from the total length traveled in the field?

File: src/field/FieldEquation.i.hh (hunk 15 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/field/FieldEquation.i.hh:67]: `op=` should definitely work for `Real3` -- otherwise we wouldn't be able to set the outgoing directions of secondaries, among other things.

Please use the celeritas data structures inside celeritas code, not the vecgeom ones. If there are features such as `operator*=` that you feel are critical to getting this online, we can add them to the array definition, or (since Real3 has no private data) you can implement free functions in a header file inside the `field` directory.

File: src/field/FieldEquation.hh (hunk 17 above)
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.hh:26]: Time (global time) will be an input for a time-dependent magnetic field case (i.e, to get field_value(time, position)).  On the other hand, it is not of course, a part of the equation, but one of required variables.

File: src/field/FieldEquation.hh [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+    //! Type aliases
+    using ode_type = OdeArray;
+    //@}
+
+  public:
+    // Construct with a magnetic field
+    CELER_FUNCTION FieldEquation(MagField& field);
+
+    // set charge if q != -eplus
+    CELER_FUNCTION void set_charge(real_type q);
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.hh:34]: Then,  FieldEquation will only work for a specific charge for a given instance.  Agree that the charge of particles should not be set via a setter of the field equation either, but should be passed by, for example, y of FieldTrackView (original idea), which will waste an additional memory of which value is likely same most of cases (i.e., electrons for EM only physics).  Alternatively I was thinking that the setter can be called with ParticleTrackView.charge() inside a field kernel.  If we separate positron kernels from electron ones, what your suggest will work more efficiently, but with less flexibility for general mixed particle cases.   What is your preference?

File: src/field/FieldEquation.hh [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+{
+  public:
+    //@{
+    //! Type aliases
+    using ode_type = OdeArray;
+    //@}
+
+  public:
+    // Construct with a magnetic field
+    CELER_FUNCTION FieldEquation(MagField& field);
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.hh:31]: Done.

File: src/field/FieldEquation.i.hh [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
+//---------------------------------------------------------------------------//
+/*!
+ * Set charge for this track and recalcuate lorentz_cof
+ */
+
+CELER_FUNCTION void FieldEquation::set_charge(real_type charge)
+{
+    charge_ = charge;
+    coeffi_ = charge_ * constants::c_light * 1.0e-8;
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.i.hh:33]: This additional conversion factor will go away in the next extension (using celeritas units everywhere in field codes).  If you don't mind, I like to leave like this (instead of introducing a temporary function, static constexpr real_type unit_conversion().

File: src/field/FieldEquation.i.hh (hunk 15 above)
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldEquation.i.hh:67]: Okay.  May I add a scale function, something like
    template<class T, std::size_t N>                                            
    CELER_FUNCTION Array<T, N> scale_array(const T a, const Array<T, N>& x)           
    {                                                                           
        Array<T, N> result;                                                     
        for (std::size_t i = 0; i != N; ++i)                                    
        {                                                                       
            result[i] = a * x[i];                                               
        }                                                                       
         return result;                                                         
    }                        

Otherwise, I will use a private function (simplified one equivalent to scale_array<real_type, 3>) inside the class (then it will be just add one more function call for nothing).

File: src/field/RungeKutta.hh (hunk 12 above)
Comment (whokion) [reply src/field/RungeKutta.hh:43]: Done.  In other places too.

File: src/field/FieldEquation.hh (hunk 20 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/field/FieldEquation.hh:34]: Oh I see -- so what we're seeing here is state data (particle charge) and parameter data (the parameterized magnetic field) getting lumped into the same class (which isn't acting like a TrackView but probably should be). There are a lot of moving pieces here, maybe we need to take a step back and think of how to better get the state-like data (position, momentum, charge, time)  to work better with the shared data (magnetic field parameterization, ODE  stencil).

File: src/field/FieldEquation.i.hh (hunk 22 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/field/FieldEquation.i.hh:33]: Ah I see. The policy we've adopted to far is for code merged into Celeritas be consistent (with the rest of the code) and correct (as far as can be initially determined), though not necessarily complete. If the units are a problem for now, could we defer the parts that use those units until the next PR?

File: src/field/FieldEquation.hh (hunk 17 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/field/FieldEquation.hh:26]: Here also I'm thinking that the OdeArray is effectively acting as a local state -- though it's only temporary and will be used to update the geometry state.

File: src/field/FieldIntegrator.i.hh (hunk 11 above)
Comment (whokion) [reply src/field/FieldIntegrator.i.hh:96]: Okay, the last is a good suggestion and codes are changed accordingly - I also considerred something similar logic in the beginning, but worried about the code redundancy. Isn't calling step(...) twice adding number of instructions and memory copies (i.e., passing arguments)?

File: src/field/FieldStepper.hh (hunk 13 above)
Comment (whokion) [reply src/field/FieldStepper.hh:51]: T will be any Runge-Kutta family stepper.  I am going to test other steppers, such as RK-Fehlberg, RK-Nystrom and etc. in the next extension, which we use the common ode_stepper method.  So, let's leave like this, but consolidate codes later (i.e, no static polymorphism) if other steppers are not as attractive as the classical-4th-order-RK.

File: src/field/FieldIntegrator.i.hh [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+        // Avoid numerous small last steps
+        condition = (h < h_threshold || curve_length >= end_curve_length);
+
+        
+        if(!condition)
+	{
+  	    if (CELER_UNLIKELY(--step_countdown == 0))
+	    {
+	        // Failed to advance, handle rare cases here
+	        break;
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/field/FieldIntegrator.i.hh:236]: Done as suggested below.

File: src/field/FieldIntegrator.i.hh (hunk 11 above)
Comment (sethrj) [reply src/field/FieldIntegrator.i.hh:96]: That's a good point, I wasn't considering the optimization angle; I was mainly thinking about the readability. Compilers are pretty smart about optimizing loops if we give them enough information.... I did create a CPU version of these three versions (plus a fourth that uses a flag to switch from setup to pre-step after the first loop): https://github.com/sethrj/testsnippets/tree/master/iteration-to-convergence . It's not going to be directly applicable to the GPU version, but the `do/while` version (the second one) has the fewest instructions even though it's the hardest to read. I guess the ideal readability plus performance would be to have a well-tested `template<class T> bool solve(T&& solver)` and define an  interface for `T` that has `step()`, `update_step()`, `converged()`, etc.... then just call `is_converged = solve(stepper);` in each of these places. The generic `solve` there could be as ugly as it needed for efficiency because it's only defined one place and we  test all the edge cases.

The mag field stuff as it is has enough small helper classes already, though, so I'd say that this iteration should focus on readability... for our next performance testing round let's try improving the loop structure.

File: src/field/FieldEquation.i.hh (hunk 15 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/field/FieldEquation.i.hh:67]: Sure thing. If it's a const transformation like this I would call it `scaled` (indicating it returns a scaled version) rather than `scale` (indicating it modifies the original).

**Comment by whokion** [comment]
@sethrj I pushed more updates except for the requested change from the mutable reference to the pointer, which I would like to do in a separate MR if you do not mind (as it involves to change in almost of all places from the top to bottom).   Thanks for reviewing!

**Comment by sethrj** [comment]
> @sethrj I pushed more updates except for the requested change from the mutable reference to the pointer, which I would like to do in a separate MR if you do not mind (as it involves to change in almost of all places from the top to bottom). Thanks for reviewing!

I don't see the advantage of merging in the current branch and then restarting the review with the new changes. There's a lot of new code in this MR, which is OK, but the use of mutable references makes it very hard to review because I have to dig through the new code each time I see a function call to find out whether the arguments are changing or not. Even if making the changes were simply a matter of superficial consistency with the rest of the code, I still don't understand why it would be better to merge the "not cleaned up" state and then clean it up in a separate commit.

Since there is so much new code here, let me suggest two options for now (maybe there are more):
- Create an "intermediate" MR with some of the lower-level helper components that can be reviewed before the higher-level solver and geometry stuff gets merged in. That'll make it easier to change in response to reviews and will help me get familiar with the code.
- For a large MR it's often good for the author to give the reviewer a sort of "in-person" introduction/walkthrough of the new component pieces. I think it be premature to document all the classes and how they relate, but we could do a zoom walkthrough if you're up for it. Maybe as the main topic for the physics meeting tomorrow?

Thanks for your patience with this process!

**Comment by whokion** [comment]
Okay, I will take the first option and will break this MR by several smaller components (may be three pieces): 
- RungeKutta stepper, Equation of motion, and a constant magnetic Field.
- Field Integration/driver (adaptive step control)
- Field Propagator and connection to the geometry 
- (integration to the transportation process as an extension to the linear transport process) 

Regarding the mutable reference, I am not much convinced whether using pointers is better for the field code as all of those are either simple variables or arrays (not complex objects or data types) - since many of them are updated together during intermediate stages of the field integration process, it is rather difficult to break code segments by the smaller pieces with a single return value (or array).  Furthermore, most of their arithmetic operations are straighforward between those arrays, for which simple c = a + b is much cleaner than *c = *a+ *b.  Anyway, I will try to break as many as smaller pieces as possible in  order to minimize multiple return arguments. 

I will remove this branch and create new ones when they are ready (may be one at a time).

**Comment by sethrj** [comment]
OK, thanks @whokion . Sorry for the trouble and the extra time and effort on your, but I believe that this process is ultimately best for the long-term stability and capability of Celeritas. I really do appreciate all the work you put into this PR and am happy to help out as much as I can with the upcoming piecemeal pull requests.

**Comment by whokion** [comment]
@sethrj No problem at all. Thanks for the pre-review which was very helpful.  Even though the first set is already ready for reviewing, I like to take this opportunity for improving/extending codes and tests more general/complete.
